[Refactor] refactor xc test setup into factory method

Leah-Xia-Microsoft opened the pull request:
## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
* Refactor XCTest setup and teardown into makeSUT factory method
* Guideline:
   - Use setUp() when something should be setup for all tests for a test class.
   - Use factory method makeSUT() to return the testing object.
   - For things that are only needed for a specific test, initialized them in the test code itself.

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What to Check
Verify that the following are valid
* All existing tests should past

## Other Information
<!-- Add any other helpful information that may be needed here. -->

Leah-Xia-Microsoft committed:
Refactor RemoteParticipantsManagerTests

Leah-Xia-Microsoft committed:
Refactor CompositeViewModelFactoryTests and other 3 test files

Leah-Xia-Microsoft committed:
Refactor tests setup for ControlBarViewModelTests

Leah-Xia-Microsoft committed:
Refactor tests setup for CallingMiddlewareHandlerTests

Leah-Xia-Microsoft committed:
Refactor tests setup for ParticipantCellViewModelTests

Leah-Xia-Microsoft committed:
Refactor tests setup for CallingViewModelTests

Leah-Xia-Microsoft committed:
Refactor test setup for InfoHeaderViewModelTests

Leah-Xia-Microsoft committed:
Refactor test setup for ParticipantsListViewModelTests

Leah-Xia-Microsoft committed:
Refactor test setup for BannerInfoTypeTests and 2 other tests

Leah-Xia-Microsoft committed:
Refactor test setup for LobbyOverlayViewModelTests

Leah-Xia-Microsoft committed:
Refactor test setup for OnHoldOverlayViewModelTests

Leah-Xia-Microsoft committed:
Refactor test setup for SetupViewModelTests

Leah-Xia-Microsoft committed:
Refactor test setup for SetupControlBarViewModelTests

Leah-Xia-Microsoft committed:
Refactor test setup for PreviewAreaViewModelTests

Leah-Xia-Microsoft committed:
Refactor test setup into factory method

AmyL219 commented on AzureCommunicationUI/sdk/AzureCommunicationUICalling/Tests/Presentation/Setup/JoiningCallActivityViewModelTests.swift:
```diff
@@ -10,16 +10,6 @@ import XCTest
 class JoiningCallActivityViewModelTests: XCTestCase {
     private var localizationProvider: LocalizationProviderProtocol!
```
Could we remove those local variable for local setup and not needed for testing? And optimize the makeSUT and setupMocking function?

vhuseinova-msft commented on AzureCommunicationUI/sdk/AzureCommunicationUICalling/Tests/Manager/AvatarManagerTests.swift:
```diff
@@ -11,16 +11,6 @@ import AzureCommunicationCommon
 class AvatarManagerTests: XCTestCase {
     var mockStoreFactory: StoreFactoryMocking!
```
Do we need `mockStoreFactory` property?
It looks like it can be used only in `makeSUT` funcs

Leah-Xia-Microsoft replied:
@vhuseinova-msft  It's also used in this test case `test_avatarManager_setRemoteParticipantViewData_when_personeDataSet_then_participantViewDataDataUpdated`

vhuseinova-msft commented on AzureCommunicationUI/sdk/AzureCommunicationUICalling/Tests/Manager/CompositeErrorManagerTests.swift:
```diff
@@ -120,4 +106,12 @@ extension CompositeErrorManagerTests {
         return handler
     }
+
+    func setupMocking() {
```
Are there any advantages of using `setupMocking` function instead of `setUp` one?

AmyL219 commented on AzureCommunicationUI/sdk/AzureCommunicationUICalling/Tests/Presentation/Calling/BannerTextViewModelTests.swift:
```diff
@@ -108,12 +98,18 @@ class BannerTextViewModelTests: XCTestCase {
     }
+
+    func setupMocking() {
+        localizationProvider = LocalizationProviderMocking()
```
where is this localizationProvider being used inside this test class?

Leah-Xia-Microsoft replied:
@AmyL219 It has been used in this test case `test_bannerTextViewModel_display_complianceBanner_from_LocalizationMocking`

Leah-Xia-Microsoft committed:
Revert "Refactor RemoteParticipantsManagerTests"

This reverts commit 179c1b943720770cbba755e7d5c5d62362ef43e8.

Leah-Xia-Microsoft committed:
Fix Refactor CompositeViewModelFactoryTests and other 3 test files

Leah-Xia-Microsoft committed:
Revert "Refactor tests setup for ControlBarViewModelTests"

This reverts commit f56cc0ec4226d9fcd2fb9bc9409c9463f90daff3.

Leah-Xia-Microsoft committed:
Fix Refactor tests setup for ControlBarViewModelTests

Leah-Xia-Microsoft committed:
Revert "Refactor tests setup for ParticipantCellViewModelTests"

This reverts commit 7c2478fdd83fc3621889ae3acb25bc0ad0fad83f.

Leah-Xia-Microsoft committed:
Revert "Refactor tests setup for CallingViewModelTests"

This reverts commit d4c5d592605229f623545321af865356158253de.

Leah-Xia-Microsoft committed:
Revert "Refactor test setup for InfoHeaderViewModelTests"

This reverts commit a12d3f0b6cfa493971938214976d6b54d256d8fe.

Leah-Xia-Microsoft committed:
Revert "Refactor test setup for ParticipantsListViewModelTests"

This reverts commit 42da76bf1989e10eb15213b9ee39a6824e6eb4eb.

Leah-Xia-Microsoft committed:
Revert "Refactor test setup for BannerInfoTypeTests and 2 other tests"

This reverts commit 150e280886cc0453dfc61545aa9c71f73df3520d.

Leah-Xia-Microsoft committed:
Revert "Refactor test setup for LobbyOverlayViewModelTests"

This reverts commit e27b67827718f58fdf21481b878dcb13b94c8837.

Leah-Xia-Microsoft committed:
Revert "Refactor test setup for OnHoldOverlayViewModelTests"

This reverts commit d2b55aee9878268298f4be413aa07bf2bc0aeb2a.

Leah-Xia-Microsoft committed:
Revert "Refactor test setup for SetupViewModelTests"

This reverts commit 0d4efe74ccc66f1edc576cf5d99e0453feadbcae.

Leah-Xia-Microsoft committed:
Revert "Refactor test setup for PreviewAreaViewModelTests"

This reverts commit f38e6ff560e399915e0584ad3f2030b5a3c7805b.

Leah-Xia-Microsoft committed:
Refactor BannerInfoTypeTests and 2 other tests

Leah-Xia-Microsoft committed:
Refactor CallingViewModelTests

Leah-Xia-Microsoft committed:
Refactor ControlBarViewModelTests

Leah-Xia-Microsoft committed:
Refactor InfoHeaderViewModelTests

Leah-Xia-Microsoft committed:
Refactor ParticipantsListViewModelTests

Leah-Xia-Microsoft committed:
Refactor PreviewAreaViewModelTests

Leah-Xia-Microsoft committed:
Fix SetupControlBarViewModelTests

Leah-Xia-Microsoft committed:
Remove unnecessary variables and rename getSUT to makeSUT

Leah-Xia-Microsoft committed:
Refactoring factoryMocking into Setup()

Leah-Xia-Microsoft committed:
Refactor tests

Leah-Xia-Microsoft committed:
Merge branch 'develop' into task/refactor-XCTest-setup-into-factory-method

vhuseinova-msft commented on AzureCommunicationUI/sdk/AzureCommunicationUICalling/Tests/Presentation/Calling/CallingViewModelTests.swift:
```diff
@@ -249,87 +294,4 @@ extension CallingViewModelTests {
                                 accessibilityProvider: accessibilityProvider,
```
I'm wondering if thie func is used anywhere?

vhuseinova-msft commented on AzureCommunicationUI/sdk/AzureCommunicationUICalling/Tests/Presentation/Calling/OnHoldOverlayViewModelTests.swift:
```diff
@@ -8,28 +8,16 @@ import XCTest
         super.setUp()
-        logger = LoggerMocking()
-        storeFactory = StoreFactoryMocking()
         localizationProvider = LocalizationProviderMocking()
```
Do we need a variable for` localizationProvider` as it's used not in the all tests?

Leah-Xia-Microsoft replied:
@vhuseinova-msft  It's used in 2/4 cases. Would you suggest initialize it in those 2 cases and pass it to a makeSUT? Feel like this is the approach that we are refactoring away from for factoryMocking.

vhuseinova-msft replied:
@Leah-Xia-Microsoft Should it be instance of `LocalizationProviderProtocol` and it's set in the tests itself (makeSut will just use it)? In this case it will follow our agreement

Leah-Xia-Microsoft replied:
We came into an agreement to keep the initialization of `localizationProvider` in the setUp(), because if we strictly following the guideline and if any of the test cases is not using this variable then it should not be initialized in the setUp(), there will be some extra work when we add or delete test cases that make certain variable not in use in every test cases anymore or vise versa.

Leah-Xia-Microsoft committed:
Merge branch 'develop' into task/refactor-XCTest-setup-into-factory-method

Leah-Xia-Microsoft committed:
Remove unused function

AmyL219 commented:
Have we updated the OneNote Unit Test guideline regarding this change?

Leah-Xia-Microsoft commented:
> Have we updated the OneNote Unit Test guideline regarding this change?

@AmyL219 Yes I've added it to the last section of the unit testing guidelines: https://microsoft.sharepoint-df.com/teams/ACSMobileUI/_layouts/OneNote.aspx?id=%2Fteams%2FACSMobileUI%2FSiteAssets%2FACS%20Mobile%20UI%20Notebook&wd=target%28Quality.one%7C8AC059AD-8D39-4D99-8FFC-CADFA2D68652%2FiOS%20Unit%20Testing%20Guidelines%7C8ABA8519-D227-3647-BEFB-BFD1966EFEB2%2F%29
onenote:https://microsoft.sharepoint-df.com/teams/ACSMobileUI/SiteAssets/ACS%20Mobile%20UI%20Notebook/Quality.one#iOS%20Unit%20Testing%20Guidelines&section-id={8AC059AD-8D39-4D99-8FFC-CADFA2D68652}&page-id={8ABA8519-D227-3647-BEFB-BFD1966EFEB2}&object-id={E5E86672-7412-6B42-86CE-0118A72971B5}&F4

Leah-Xia-Microsoft committed:
Refactor makeSUTLocalizationMocking

emlynmac reviewed (approved):
LGTM
